rq/worker.py

```diff
@@ -1378,3 +1389,213 @@ class RandomWorker(Worker):
 
     def reorder_queues(self, reference_queue):
         shuffle(self._ordered_queues)
+
+
+class ThreadPoolWorker(Worker):
+    death_penalty_class = TimerDeathPenalty  # type: ignore
+
+    def __init__(self, *args, **kwargs):
+        self.threadpool_size = kwargs.pop('pool_size', self.default_pool_size)
+        self.executor = ThreadPoolExecutor(max_workers=self.threadpool_size, thread_name_prefix="rq_workers_")
+        self._idle_threads = self.threadpool_size
+        self._lock = threading.Lock()
+        self._current_jobs: List[Tuple['Job', 'Future']] = []  # type: ignore
```

Comment:
`current_jobs` is also something that we should bring into RQ proper. Mind making a PR for this so we can merge this into master?

Reply:
Sure, but the current signature won't make sense. Do you want just the attribute and an empty array as a placeholder?

Reply:
So `worker.set_current_job_id()` (and other methods saving job ID) would also save the job ID to `worker._current_jobs`.

The existing `worket.get_current_job()` method would simply take the first job from the array so it remains backward compatible.

Reply:
Understood. This is not how this method is being used in this case. We may adapt and change the logic for the `setter` and `getter` of the current job on the current worker (there's the question whether this would be the `id` or the job instance). So regardless of whether this is implemented in the current worker or not, the current structure would still be necessary (and overlapping) for this worker (and it seems that for `GeventWorker` as well).

I understand the direction of having a more generic `Worker` class so that new implementations would be easier, but I'd like to propose a "step-by-step" approach where after implementing this (and others) workers and eventually reviewing the current worker mechanics, we may implement improvements - otherwise we are going to be concerned about (1) backwards compatibility (2) compatibility between workers (3) usefulness to other workers - it seems rather early to consider all of this for optimization sake.

Reply:
But I think `worker.get_jobs()` needs to be a standard method, since this will be used by monitoring tools like django-rq - (they need to know what job(s) are being worked on by a worker).

Reply:
For reutilizing components, not sure this makes much of a difference because the different types each concurrency model uses. Shouldn't make much difference to include that though, except the `getter` method for the current worker would be getting data from an internal object vs getting "fresh" data from the server.

Since those APIs don't exists, we might want to have the unified API done through the main `RQ` class (a proposal was shared [here](https://github.com/lowercase00/rq/blob/feature/app/rq/main.md)), this would avoid having to worry about the analytics on the `Workers` themselves and allow for more flexibility. To make new implementations predictable, we might want to write an interface to make sure new workers follow the right implementation. But all of these seem to be a different feature set.

Let me know how you want to proceed with this.

Reply:
Hi, looks good, but there are a things that I'd like to change to make it more OOP like. For example I'd switch
```python
w = rq.get_worker("myworker")
# Kill a worker
rq.kill_worker("myworker")
```

To this:
```python
w = rq.get_worker("myworker")
# Kill a worker
w.kill()
```

So next what we can do is define the interface for each `Queue`, `Worker` and `Job`. Out of all these, the `Worker` interface is most relevant for this PR.

Reply:
```python
w = rq.get_worker("myworker")

# Here, the worker has a `kill` method.
w.kill()
```

```python
rq = RQ(...)

# Here you're making all interactions available through a general `RQ` class
rq.kill_worker("myworker")
```